libs/entitlement/src/main/java/org/elasticsearch/entitlement/runtime/policy/FileAccessTree.java

```diff
@@ -30,17 +30,87 @@
 
 public final class FileAccessTree {
 
+    record ExclusiveFileEntitlement(String componentName, String moduleName, FilesEntitlement filesEntitlement) {}
+
+    record ExclusivePath(String componentName, String moduleName, String path) {}
+
+    static List<ExclusivePath> buildExclusivePathList(List<ExclusiveFileEntitlement> exclusiveFileEntitlements, PathLookup pathLookup) {
+        List<ExclusivePath> exclusivePaths = new ArrayList<>();
+        for (ExclusiveFileEntitlement efe : exclusiveFileEntitlements) {
+            for (FilesEntitlement.FileData fd : efe.filesEntitlement().filesData()) {
+                if (fd.exclusive()) {
+                    List<Path> paths = fd.resolvePaths(pathLookup).toList();
+                    for (Path path : paths) {
+                        exclusivePaths.add(new ExclusivePath(efe.componentName(), efe.moduleName(), normalizePath(path)));
+                    }
+                }
+            }
+        }
+        exclusivePaths.sort((ep1, ep2) -> PATH_ORDER.compare(ep1.path(), ep2.path()));
+        return exclusivePaths;
+    }
+
+    static void validateExclusivePaths(List<ExclusivePath> exclusivePaths) {
+        if (exclusivePaths.isEmpty() == false) {
+            ExclusivePath currentExclusivePath = exclusivePaths.get(0);
+            for (int i = 1; i < exclusivePaths.size(); ++i) {
+                ExclusivePath nextPath = exclusivePaths.get(i);
+                if (currentExclusivePath.path().equals(nextPath.path) || isParent(currentExclusivePath.path(), nextPath.path())) {
```

Comment:
Ah, this is pretty subtle.

Assuming the incoming list of paths has been sorted by `PATH_ORDER`, that does not ensure that a child is always immediately after its parent; but it _does_ ensure that any list of paths containing one or more parent-child pairs will have at least one of those pairs adjacent, and that's enough!

For example:

```
  /a
  /a/b
  /a/c
```

The child `/a/c` is not next to its parent `/a`, but for the sake of validation, that's ok, because to have something else (like `/a/b`) interleave, that must itself be a child (or descendant) of `/a`, and so the validation will fail on that.

I'm 95% sure this reasoning works and we're not missing a case.

Reply:
Should be path order and since that moves file separators to be the first character this should work. This works similarly to pruning and *I think* has the same guarantees here. Thanks for checking it :)